mne/io/snirf/_snirf.py

```diff
@@ -0,0 +1,202 @@
+# Authors: Robert Luke <mail@robertluke.net>
+#
+# License: BSD (3-clause)
+
+import re as re
+import numpy as np
+
+from ..base import BaseRaw
+from ..meas_info import create_info
+from ...annotations import Annotations
+from ...utils import logger, verbose, fill_doc, warn, requires_h5py
+
+
+@requires_h5py
+@fill_doc
+def read_raw_snirf(fname, preload=False, verbose=None):
+    """Reader for a SNIRF fNIRS recording.
+
+    .. note:: This reader supports the .snirf file type only,
+              not the .jnirs version.
+              Currently only support for continuous wave SNIRF data is
+              provided.
+
+    Parameters
+    ----------
+    fname : str
+        Path to the SNIRF data file.
+    %(preload)s
+    %(verbose)s
+
+    Returns
+    -------
+    raw : instance of RawSNIRF
+        A Raw object containing SNIRF data.
+
+    See Also
+    --------
+    mne.io.Raw : Documentation of attribute and methods.
+    """
+    return RawSNIRF(fname, preload, verbose)
+
+
+def _open(fname):
+    return open(fname, 'r', encoding='latin-1')
+
+
+@requires_h5py
+@fill_doc
+class RawSNIRF(BaseRaw):
+    """Raw object from a SNIRF fNIRS file.
+
+    Parameters
+    ----------
+    fname : str
+        Path to the SNIRF data file.
+    %(preload)s
+    %(verbose)s
+
+    See Also
+    --------
+    mne.io.Raw : Documentation of attribute and methods.
+    """
+
+    @verbose
+    def __init__(self, fname, preload=False, verbose=None):
+        from ...externals.pymatreader.utils import _import_h5py
+        h5py = _import_h5py()
+        logger.info('Loading %s' % fname)
+
+        with h5py.File(fname, 'r') as dat:
+
+            if 'data2' in dat['nirs'].keys():
+                warn("File contains multiple recordings. "
+                     "MNE does not support this feature.")
+
+            if np.array(dat.get('nirs/data1/measurementList1/dataType')) != 1:
+                warn("File does not contain continuous wave data. "
+                     "MNE only supports reading "
+                     "continuous wave amplitude SNIRF files.")
+
+            data = np.array(dat.get('/nirs/data1/dataTimeSeries')).T
```

Comment:
This will read the entire thing, right (bad memory use and efficiency)? Can you just get the information like:
```
last_samps = dat.get('/nirs/data1/dataTimeSeries').shape[0]
```
Assuming the `.get` does not actually pull all the data until you `np.array` cast it. If it does, I have to imagine there is some way to avoid this in h5py and just get the element shape.